### ISSUE TITLE
Cannot Compile OpenMPI3 with Cuda Support

### ISSUE THREAD

**Issue by JusticeForMikeBrown** [open-mpi/ompi#5510]
Hi

I am trying to compile OpenMPI 3 with Cuda support on CentOS 7.4 with P100s on Dell PowerEdge.

We do this in [Spack](spack.io).

The command-line compilation options being used:

```
==> '/pbtech_mounts/softlib001/apps/EL7/spack/var/spack/stage/openmpi-3.0.0-p6nylowvg5mycctexaxlmg3tlqlzvfyb/openmpi-3.0.0/configure' '--prefix=/pbtech_mounts/softlib001/apps/EL7/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/openmpi-3.0.0-p6nylowvg5mycctexaxlmg3tlqlzvfyb' '--enable-shared' '--enable-static' '--enable-mpi-cxx' '--with-zlib=/pbtech_mounts/softlib001/apps/EL7/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/zlib-1.2.11-64vg6e4evdrlqgx7iicwhu2hs7lv6gpz' '--without-psm' '--without-psm2' '--with-verbs' '--without-mxm' '--without-ucx' '--without-libfabric' '--without-alps' '--without-lsf' '--without-tm' '--without-slurm' '--without-sge' '--without-loadleveler' '--without-pmi' '--disable-memchecker' '--with-hwloc=/pbtech_mounts/softlib001/apps/EL7/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/hwloc-1.11.9-mis4vfsuso2at2i2nwgyylcmj3iybpdq' '--disable-java' '--disable-mpi-java' '--enable-dlopen' '--with-cuda=/pbtech_mounts/softlib001/apps/EL7/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/cuda-8.0.61-oehfcuf3do4bwagkarjurizvcsrl7y2q'
configure: WARNING: unrecognized options: --without-loadleveler
```

This normally works but now I see this error:

```
make[2]: Leaving directory `/tmp/root/spack-stage/spack-stage-pzqhmj/openmpi-3.0.0/opal/mca/shmem/posix'
Making all in mca/shmem/sysv
make[2]: Entering directory `/tmp/root/spack-stage/spack-stage-pzqhmj/openmpi-3.0.0/opal/mca/shmem/sysv'
  CC       shmem_sysv_component.lo
  CC       shmem_sysv_module.lo
  CCLD     libmca_shmem_sysv.la
make[2]: Leaving directory `/tmp/root/spack-stage/spack-stage-pzqhmj/openmpi-3.0.0/opal/mca/shmem/sysv'
Making all in mca/timer/linux
make[2]: Entering directory `/tmp/root/spack-stage/spack-stage-pzqhmj/openmpi-3.0.0/opal/mca/timer/linux'
  CC       timer_linux_component.lo
  CCLD     libmca_timer_linux.la
make[2]: Leaving directory `/tmp/root/spack-stage/spack-stage-pzqhmj/openmpi-3.0.0/opal/mca/timer/linux'
Making all in .
make[2]: Entering directory `/tmp/root/spack-stage/spack-stage-pzqhmj/openmpi-3.0.0/opal'
  CC       class/opal_bitmap.lo
  CC       class/opal_free_list.lo
  CC       class/opal_hash_table.lo
  CC       class/opal_hotel.lo
  CC       class/opal_tree.lo
  CC       class/opal_list.lo
  CC       class/opal_object.lo
  CC       class/opal_graph.lo
  CC       class/opal_lifo.lo
  CC       class/opal_fifo.lo
  CC       class/opal_pointer_array.lo
  CC       class/opal_value_array.lo
  CC       class/opal_ring_buffer.lo
  CC       class/opal_rb_tree.lo
  CC       memoryhooks/memory.lo
  CC       runtime/opal_progress.lo
  CC       runtime/opal_finalize.lo
  CC       runtime/opal_init.lo
  CC       runtime/opal_params.lo
  CC       runtime/opal_cr.lo
  CC       runtime/opal_info_support.lo
  CC       runtime/opal_progress_threads.lo
  CC       threads/condition.lo
  CC       threads/mutex.lo
  CC       threads/thread.lo
  CC       threads/wait_sync.lo
  CC       dss/dss_compare.lo
  CC       dss/dss_copy.lo
  CC       dss/dss_internal_functions.lo
  CC       dss/dss_dump.lo
  CC       dss/dss_load_unload.lo
  CC       dss/dss_lookup.lo
  CC       dss/dss_pack.lo
  CC       dss/dss_peek.lo
  CC       dss/dss_register.lo
  CC       dss/dss_print.lo
  CC       dss/dss_open_close.lo
  CC       dss/dss_unpack.lo
  CCLD     libopen-pal.la
/usr/bin/ld: cannot find -lnvidia-ml
collect2: error: ld returned 1 exit status
make[2]: *** [libopen-pal.la] Error 1
make[2]: Leaving directory `/tmp/root/spack-stage/spack-stage-pzqhmj/openmpi-3.0.0/opal'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/tmp/root/spack-stage/spack-stage-pzqhmj/openmpi-3.0.0/opal'
make: *** [all-recursive] Error 1
```

Cuda does exist:

```
[root@node177 ~]# ls -ltr /pbtech_mounts/softlib001/apps/EL7/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/cuda-8.0.61-oehfcuf3do4bwagkarjurizvcsrl7y2q
total 52
drwxr-xr-x  7 root root 4096 Apr 25  2017 libnvvp
drwxr-xr-x  7 root root   80 Apr 25  2017 nvvm
drwxr-xr-x  3 root root   20 Apr 25  2017 nvml
drwxr-xr-x  5 root root   83 Apr 25  2017 extras
drwxr-xr-x  5 root root 8192 Apr 25  2017 include
drwxr-xr-x  8 root root 4096 Apr 25  2017 libnsight
drwxr-xr-x  2 root root   42 Apr 25  2017 tools
drwxr-xr-x  3 root root   16 Apr 25  2017 share
drwxr-xr-x  5 root root 4096 Apr 25  2017 jre
drwxr-xr-x  3 root root 4096 Apr 25  2017 lib64
drwxr-xr-x  5 root root   37 Apr 25  2017 doc
-rw-r--r--  1 root root   20 Apr 25  2017 version.txt
drwxr-xr-x  2 root root 4096 Apr 25  2017 src
drwxr-xr-x 11 root root 4096 Apr 25  2017 samples
drwxr-xr-x  2 root root 4096 Apr 25  2017 pkgconfig
drwxr-xr-x  3 root root 4096 Apr 25  2017 bin
```
Drivers are present:

```
[root@node177 ~]# nvidia-smi 
Fri Aug  3 08:29:19 2018       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 384.81                 Driver Version: 384.81                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla P100-PCIE...  Off  | 00000000:03:00.0 Off |                    0 |
| N/A   26C    P0    31W / 250W |  15747MiB / 16276MiB |      7%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla P100-PCIE...  Off  | 00000000:04:00.0 Off |                    0 |
| N/A   26C    P0    34W / 250W |  15747MiB / 16276MiB |      3%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla P100-PCIE...  Off  | 00000000:82:00.0 Off |                    0 |
| N/A   26C    P0    30W / 250W |  15747MiB / 16276MiB |      9%      Default |
+-------------------------------+----------------------+----------------------+
|   3  Tesla P100-PCIE...  Off  | 00000000:83:00.0 Off |                    0 |
| N/A   29C    P0    30W / 250W |  15747MiB / 16276MiB |      3%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0    105406      C   ...dcxakaqdoin2ugxnr/bin/relion_refine_mpi 15737MiB |
|    1    105406      C   ...dcxakaqdoin2ugxnr/bin/relion_refine_mpi 15737MiB |
|    2    105407      C   ...dcxakaqdoin2ugxnr/bin/relion_refine_mpi 15737MiB |
|    3    105407      C   ...dcxakaqdoin2ugxnr/bin/relion_refine_mpi 15737MiB |
+-----------------------------------------------------------------------------+
```

**/etc/ld.so.conf.d/nvidia-lib64.conf**

```
/usr/lib64/nvidia
```

```
[root@node177 ~]# ls -ltr /usr/lib64/nvidia | grep nvidia-ml
-rwxr-xr-x 1 root root  1304256 Sep  2  2017 libnvidia-ml.so.384.81
lrwxrwxrwx 1 root root       22 Oct 17  2017 libnvidia-ml.so.1 -> libnvidia-ml.so.384.81
lrwxrwxrwx 1 root root       22 Oct 17  2017 libnvidia-ml.so -> libnvidia-ml.so.384.81
```

Could you please help as I am unsure why this isn't working.

[spack-build.txt](https://github.com/open-mpi/ompi/files/2257332/spack-build.txt)

**Comment by jsquyres** [comment]
@Akshay-Venkatesh Can you answer @JusticeForMikeBrown's question?  Thanks.

**Comment by JusticeForMikeBrown** [comment]
Thank you @jsquyres @Akshay-Venkatesh

**Comment by JusticeForMikeBrown** [comment]
Hi

I was able to compile on a server with older Nvidia driver.

384.81 unsupported?

**Comment by Akshay-Venkatesh** [comment]
@JusticeForMikeBrown 
The error suggests that build system can't find -lnvidia-ml. You're able to compile on your server because the build system is likely able to find it in the root lib search path. Are you building on a node177? Also want to point at that the location `/pbtech_mounts/softlib001/apps/EL7/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/cuda-8.0.61-oehfcuf3do4bwagkarjurizvcsrl7y2q/lib64` doesn't point to nvml. As you're seeing, it's `/usr/lib64` that has it.  

I would make sure that the build system can find the library first. Lmk what you find.

@jsquyres 
Only hwloc component depends on -lnvidia-ml (NVML). Can you also @  hwloc cuda component developer here in case I'm missing something obvious?

**Comment by jsquyres** [comment]
FTI @bgoglin

**Comment by bgoglin** [comment]
OMPI 3.0.1 has a fix that unconditionally disables nvml in hwloc to avoid such issues.

**Comment by jsquyres** [comment]
Looks like this was resolved in Open MPI v3.0.1.  Closing the issue.
